src/ast/passes/semantic_analyser.cpp

```diff
@@ -1914,6 +1915,127 @@ void SemanticAnalyser::visit(While &while_block)
   loop_depth_--;
 }
 
+void SemanticAnalyser::visit(For &f)
+{
+  bool error = false;
+
+  if (!bpftrace_.feature_->has_helper_for_each_map_elem()) {
+    LOG(ERROR, f.loc, err_)
+        << "Missing required kernel feature: for_each_map_elem";
+    error = true;
+  }
+
+  /*
+   * For loops are implemented using the bpf_for_each_map_elem helper function,
+   * which requires them to be rewritten into a callback style.
+   *
+   * In this first implementation, we do not allow any variables to be shared
+   * between the main probe and the loop's body. Maps are global so are not a
+   * problem.
+   *
+   * Pseudo code for the transformation we apply:
+   *
+   * Before:
+   *     PROBE {
+   *       @map[0] = 1;
+   *       for ($kv : @map) {
+   *         [LOOP BODY]
+   *       }
+   *     }
+   *
+   * After:
+   *     PROBE {
+   *       @map[0] = 1;
+   *       bpf_for_each_map_elem(@map, &map_for_each_cb, 0, 0);
+   *     }
+   *     long map_for_each_cb(bpf_map *map,
+   *                          const void *key,
+   *                          void *value,
+   *                          void *ctx) {
+   *       $kv = ((uint64)key, (uint64)value);
+   *       [LOOP BODY]
+   *     }
+   */
+
+  // Validate decl
+  const auto &decl_name = f.decl->ident;
+  if (variable_val_[probe_].find(decl_name) != variable_val_[probe_].end()) {
+    LOG(ERROR, f.decl->loc, err_)
+        << "Loop declaration shadows existing variable: " + decl_name;
+    error = true;
+  }
+
+  // Validate expr
+  if (!f.expr->is_map) {
+    LOG(ERROR, f.expr->loc, err_) << "Loop expression must be a map";
+    error = true;
+  }
+  Map &map = static_cast<Map &>(*f.expr);
+
+  // Validate body
+  CollectNodes<Variable> vars_referenced;
+  for (auto *stmt : *f.stmts) {
+    vars_referenced.run(*stmt);
+  }
+  for (const Variable &var : vars_referenced.nodes()) {
+    if (variable_val_[probe_].find(var.ident) != variable_val_[probe_].end()) {
+      LOG(ERROR, var.loc, err_) << "Variables defined outside of a for-loop "
+                                   "can not be brought into the loop's scope";
+      error = true;
+    }
+  }
+
+  // This could be relaxed in the future:
+  CollectNodes<Jump> jumps;
+  for (auto *stmt : *f.stmts) {
+    jumps.run(*stmt);
+  }
+  for (const Jump &n : jumps.nodes()) {
+    LOG(ERROR, n.loc, err_)
+        << "'" << opstr(n) << "' statement is not allowed in a for loop";
+    error = true;
+  }
+
+  if (error)
+    return;
+
+  map.skip_key_validation = true;
+  map.accept(*this);
+
+  // Iterating over a map provides a tuple: (map_key, map_val)
+  auto *mapkey = get_map_key_type(map);
+  auto *mapval = get_map_type(map);
+
+  if (mapkey && mapkey->args_.size() == 0) {
+    LOG(ERROR, map.loc, err_)
+        << "Maps used as for-loop expressions must have keys to iterate over";
+  }
+
+  if (mapval && mapkey) {
```

Comment:
It can happen before type-propagation is complete if a map is defined after being used in a for-loop, e.g.:

`BEGIN { for ($kv : @map) {} } END { @map[0] = 1; }`

`@map` doesn't exist on the first semantic analyser pass.